vm/devices/storage/disk_nvme/nvme_driver/src/namespace.rs

```diff
@@ -499,6 +520,70 @@ impl Namespace {
 
         Ok(())
     }
+
+    /// Return Namespace ID.
+    pub fn nsid(&self) -> u32 {
+        self.nsid
+    }
+
+    /// Save namespace object data for servicing.
+    pub fn save(&self) -> anyhow::Result<SavedNamespaceData> {
+        let id = self.state.identify.lock();
+        let mut save_data = SavedNamespaceData {
+            nsid: self.nsid,
+            block_count: self.state.block_count.load(Ordering::Relaxed),
+            identify_ns: [0; 4096],
+        };
+        save_data.identify_ns.copy_from_slice(id.as_bytes());
+        Ok(save_data)
+    }
+
+    /// Restore namespace object data after servicing.
+    pub(super) fn restore(
+        driver: &VmTaskDriver,
+        admin: Arc<Issuer>,
+        rescan_event: Arc<event_listener::Event>,
+        identify_ctrl: Arc<spec::IdentifyController>,
+        io_issuers: &Arc<IoIssuers>,
+        device_id: &str,
+        identify_ns: &[u8; 4096],
+        saved_state: &SavedNamespaceData,
+    ) -> Result<Self, NamespaceError> {
+        let nsid = saved_state.nsid;
+        let identify = nvm::IdentifyNamespace::read_from_prefix(identify_ns)
+            .unwrap_or(nvm::IdentifyNamespace::new_zeroed());
+
+        let (mut ctx, cancel_rescan) = CancelContext::new().with_cancel();
+        let this = Namespace::new_from_identify(
+            identify_ctrl.clone(),
+            io_issuers,
+            cancel_rescan,
+            nsid,
+            identify,
+        )?;
+
+        // Spawn a task, but detach is so that it doesn't get dropped while NVMe
```

Comment:
Don't copy/paste this code. Can this not be in `new_from_identify`?

Reply:
Hmmm I think I had issues with async/non-async, let me check again

Reply:
Moved everything into `new_from_identify`